### PR TITLE
Prevent unnecessary browsers reflows and hide flash of unstyled content

### PR DESCRIPTION
**Pull request by jleider**
This PR addresses #472, where using jQuery 3+ with FooTable would cause a tremendous number of repaints and reflows in the browser. The concept behind this change is to remove the table from the DOM rendering flow while FooTable performs its draw operations and then placing it back once the table has been redrawn. During operations where height is generally not affected such as sorts, FooTable will also set the height directly on the table element before changing visibility to prevent reflows. 

Additionally, I have added a convenience class that users may add to their table html element which hides the table until the FooTable has been initialized. The goal again here is to hide FOUC until FooTable has initialized.

### CODE REVIEW

**Comment by steveush** [comment]
Thanks @jleider! I'll test this out and merge if everything looks good.

**Comment by jleider** [comment]
@steveush I just updated this branch again with the help of @mrdziuban, we came up with a better solution which prevents a flash of hidden content when performing draw operations on an initialized table.

**Comment by mrdziuban** [comment]
Here are some gifs to demonstrate the changes.

##### Before

slow_paint

##### With reflow improvements, but with FOUC

faster_paint

##### With reflow improvements and hiding of FOUC via CSS class

better_paint

**Comment by steveush** [comment]
Hey guys,

I haven't merged this just yet as there seems to be an issue with the pagination buttons layout due to the table being either detached from the body or hidden using `.hide()` at the time they are created. The below shows the normal layout as seen in the live [Ajax example](http://fooplugins.github.io/FooTable/docs/examples/advanced/ajax.html).

image

Then when testing the PR I get the following:

image

There is also some code cleanup to do:

* The CSS class `table.footable-hide-fouc` doesn't seem to be used at all and can be removed entirely.
* The following code block:

```javascript
// Copy `self.$el` with no event handlers
var $elCopy = $(self.$el[0].outerHTML);

// Insert the copy at the same location as `self.$el`
self.$el.prev().length === 0
	? self.$el.parent().prepend($elCopy)
	: self.$el.prev().after($elCopy);
```

Can probably be refactored down to a single line.

```javascript
// clone the current table and insert it into the orginals place
var $elCopy = self.$el.clone().insertBefore(self.$el);
```

jQuery's `.clone()` does not copy events and data by default, it's purely deep copy of the elements, so it can be used here. Then the `.insertBefore()` method lets us insert the clone into the DOM just before the original table, effectively taking it's place before we detach it.

Thanks

**Comment by mrdziuban** [comment]
@steveush thanks for getting back to us. I refactored the `$elCopy` stuff into a single line.

For the `table.footable-hide-fouc` class, the intent was to provide it as a utility class that users can add to their `table` elements. Adding the class ensures that the content will not flash while the table is initializing.

As for the pagination layout, I'm having trouble reproducing the problem you're seeing. I created two jsfiddles to show that it's working:

- Using FooTable without Bootstrap: https://jsfiddle.net/2kco4eLe/2/
- Using FooTable with Bootstrap: https://jsfiddle.net/frbz73k1/1/

Can you let us know how you're able to reproduce it?

**Comment by steveush** [comment]
@mrdziuban no no, thank you for contributing and sorry I haven't had a chance to dig into this and figure out the paging stuff on my side yet, just been really busy the last few weeks.

As for reproducing it I literally just cloned the PR to my local machine and without making any changes I started going through the advanced examples, as if there's anything really wrong they would show it first. I think the first one I opened and tested was the Ajax example and noticed the pagination was laid out as in my screenshot above.

That was all I did to get the issue on my side, I am using Chrome Version 54.0.2840.99 m on Windows 10, I haven't checked if the issue is reproducible in FF or IE/Edge yet as Chrome is my primary and generally what I code against.

**Comment by mrdziuban** [comment]
No problem! I was just able to reproduce the issue by using jQuery 1.x and 2.x. I'm still digging into what makes it work in jQuery 3.x, but here's a jsfiddle demonstrating the issue: https://jsfiddle.net/ha61ez43/1/.

**Comment by mrdziuban** [comment]
@steveush we figured out that the cause was a change in jQuery 3 in the `hide` and `show` functions. When calling `show` in earlier versions, jQuery would try to intelligently pick the default `display` property for the given element type, but in jQuery 3, it just removes the `display` property. The CSS was broken in jQuery 1.x and 2.x because the inline style of `display: list-item` on certain elements was overriding the intended `display: inline` property declared in the CSS file.

As you can see in the last commit I pushed, the effected elements were having `hide` and `show` called on them. Changing those calls to manually modify the `display` property instead fixes the issue. I tested the fix with all 3 versions of jQuery and everything looked good.

**Comment by steveush** [comment]
@mrdziuban nice one :) I'll pull the latest and give it a quick test run and if it's all good I'll merge it into develop tonight.

**Comment by steveush** [comment]
@mrdziuban and @jleider thanks guys, I've run through all the examples in the usual browsers and everything seems to be good so I've merged this PR into develop and it will be included in the next release.
